autoload/neomake/makers/ft/python.vim

```diff
@@ -296,3 +296,45 @@ function! neomake#makers#ft#python#py3kwarn() abort
         \ 'errorformat': '%W%f:%l:%c: %m',
         \ }
 endfunction
+
+" Find out the python version to be used in the current buffer (script)
+" by checking the shebang. If no shebangis found or it does not conatain
+" an explicit version number, use the default interpreter.
+" If the shebang changed since the last write, unset the buffer specific
+" settings, then set the new values.
+function! neomake#makers#ft#python#SelectAlternatives() abort
+
+    if !exists('s:default_version')
+        " Get defult python executable version as [major, minor, revision]
+        let s:default_version = split(matchstr(system('python --version'), '[0-9\.]\+'), '\')
+    endif
+
+    " Scan the shebang
+    let l:shebang = matchlist(getline(1), '^#!.*python\([0-9.]*\)')
+
+    if empty(l:shebang) || l:shebang[1] == ''
+        " Use the major number from s:default_version
+        " Ip $ the shebang does not contain one or is absent
+        let l:pyversion = s:default_version[0]
+    else
+        let l:pyversion = l:shebang[1]
```

Comment:
Let see:
only on the first call, the function evaluates `python --version` and stores the version number in s:pyversion (in my system this would be ['2, '7', '13'] right now)

then scan the first line the regex can only match if the line begins wit '#!' and contains python as last word, possibly followed by an explicit version number. The possible outcomes are:
#!<anything> <anythingelse>  ->  []
#!<anything> python  ->  ['#!<anything> python', '', '', '', '', '', '', '', '', '']
#!<anything> python<version>  ->  ['#!<anything> python<version>', '<version>', '', '', '', '', '', '', '', '']

Here <anything> should be '/usr/bin/env', but I made it more generic just in case of strange installations, <version> can be any number with or without dots e.g.: 2, 2.7, 3, 3.6.

If empty(l:shebang), the regex did not match, the second part of the condition at line 315 is not evaluated and the default version is used.
if ! empty(l:shebang) the regex matched, the requested interpreter is 'python<version>' and l:shebang[1] == '' is checked: the latter is only true if no explicit version number is present and again the default version is used.
it is not possible for both parts of the condition to be true and if line 320 is reached, the name of the interpreter contains a version number and I would use just that since it should be the script author responsibility to set the correct shebang (note that in the default I only use the major number, while here more specific requests are allowed).

There is the possibility that the requested interpreter is not present in the system (or evenany python at all, but this feature is meant to be an opt-in, so it should not be an issue.

It is good that I checked, since I spotted a bug that would trigger only if I were to configure options for 'key' == '2'